### ISSUE TITLE
POLARS_FMT_MAX_ROWS

### ISSUE THREAD

**Issue by n8henrie** [pola-rs/polars#6825]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://crates.io/crates/polars) of Polars.

### Issue description

I've been trying to figure out how to remove the row limit when displaying a dataframe.

Looking at the source, I saw that `FMT_MAX_ROWS` is a `usize`, so I set `std::env::set_var("POLARS_FMT_MAX_ROWS, usize::MAX.to_string())`. I was surprised to find that this didn't work -- still truncates at 8.

Looking deeper, there is a closure that casts `limit` to `i64`, which panics and `unwrap_or`s to the lower value:

https://github.com/pola-rs/polars/blob/c6a3df09beca709923029635f563a305251f5d06/polars/polars-core/src/fmt.rs#L75

I'd be happy to submit a (obviously very small) PR if this is not intentional behavior.

### Reproducible example

```rust
use polars::prelude::*;

fn main() {
    std::env::set_var("POLARS_FMT_MAX_ROWS", usize::MAX.to_string());
    let df = df! ["foo" => 0..100].unwrap();
    println!("{df}");
}
```

```console
$ cargo run -q
shape: (100, 1)
┌─────┐
│ foo │
│ --- │
│ i32 │
╞═════╡
│ 0   │
│ 1   │
│ 2   │
│ 3   │
│ ... │
│ 96  │
│ 97  │
│ 98  │
│ 99  │
└─────┘
```

### Expected behavior

Should display up to `usize::MAX` rows, or perhaps `limit` should be `i64`, but probably not both.

### Installed versions

0.27.2

**Comment by slonik-az** [comment]
> I've been trying to figure out how to remove the row limit when displaying a dataframe.

Setting POLARS_FMT_MAX_ROWS to "-1" or any other negative number prints all rows in both Rust and Python. For discussion why it is negative number rather than 0 that removes truncation see #5047 and #5046.
This explains why POLARS_FMT_MAX_ROWS is parsed as `i64` in the `fmt.rs` code you referred to. And it has to work for python (and Javascript) as well. 

It does not seem to be a bug here. We might need to document this vehavior better, though.

**Comment by ritchie46** [comment]
We could do an overflow check.

**Comment by n8henrie** [comment]
Thanks for the explanation. I had *expected* a flag value such as `-1` to serve this purpose but didn't find it documented, which is why I went to the source. I obviously didn't understand what that closure was doing (looking for `-1` specifically vs `< 0` might have been more obvious that this was intentional).

If not a bug, I'm not sure what else to call it when `LIMIT` and `limit` are both `usize`, but requesting `usize::MAX` rows gets me back 8 ¯\\\_(ツ)_/¯.

I would think less surprising behavior could be obtained by:

1. Documentation
2. Either make the closure take a `i128`, reducing overflow concerns, or emit a warning in case of overflow, or -- what I think would be optimal -- to match on `"-1"` prior to parsing, making this a truly a "flag" value (as opposed to `-2` or `-478` rows, which are probably user error), and otherwise parse as `usize`. I don't see any reason to include `-2`-`i64::MIN` as valid values.

Happy to contribute to both of these if they seem reasonable!

**Comment by slonik-az** [comment]
> Looking at the source, I saw that `FMT_MAX_ROWS` is a `usize`

This is incorrect. `FMT_MAX_ROWS` is a `const` of type `&str` that simply names the environment variable.

https://github.com/pola-rs/polars/blob/264cc0c5c284d49e5249425aab90e0a062c5fca5/polars/polars-core/src/config.rs#L2-L3

Polars never specifies the dynamic range of the numbers represented by the environment variables like `POLARS_FMT_MAX_ROWS`. And in practice, `isize:max` is as good as `usize::max` because you would run out the __address space__ well before creating a DataFrame with `isize::max` rows or columns.

But I think we should add to polars crate documention that setting these environment variables to negative numbers disable rows/columns truncation.

**Comment by n8henrie** [comment]
> This is incorrect. FMT_MAX_ROWS is a const of type &str that simply names the environment variable.

Clearly I misspoke, my apologies for being sloppy. Hopefully you were able to understand what I meant, especially given my more accurate phrasing in my more recent post, where I specify "LIMIT and limit are both usize".

> And in practice, isize:max is as good as usize::max

Well it would be, if it trying to specify the limit as `usize::MAX` didn't overflow and surprisingly revert to 8. I'm not trying to argue that `usize` is better than `isize` (or `i64` in this case, right?), just that having all of the relevant variables be `usize` (other than `FMT_MAX_ROWS` as you've pointed out) but then overflow at `i64` in that closure was surprising.

> setting these environment variables to negative numbers

Reviewing the prior linked threads (thank you), it seems like there was some nuanced conversation about `0` not being appropriate because it would lead to *unexpected behavior* if someone "asked for zero apples." Does `-4` mean something different than `-1` in this case?

If not, what is the downside to something like:

```rust
const LIMIT: usize = 25;

fn main() {
    let a = [1, 2, 3];
    let limit: usize = match std::env::var("POLARS_FMT_MAX_ROWS").as_deref() {
        Ok("-1") => a.len(),
        Ok(n) => n
            .parse()
            .unwrap_or_else(|_| panic!("Unable to parse {n} as usize")),
        _ => LIMIT,
    };
    dbg!(limit);
}
```

Anyway, I agree that a documentation fix would be adequate -- I'm happy to work on that if it would be helpful.

**Comment by slonik-az** [comment]
Exactly like you said my original PR was using `usize` for the type and `0` as "unlimited" value. But after the discussion I was persuaded to  change it to signed integer type and negative numbers meaning unlimited. "zero apples" was a deal breaker 😄

**Comment by alexander-beedie** [comment]
Lol... as the man wanting "zero apples", I will briefly mention that I did go back into the formatting/layout code and [fix the issue](https://github.com/pola-rs/polars/pull/5281) whereby you couldn't print a table with zero rows ;)
